de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,537 @@
+                + referencePoint);
+            return;
+        }
+
+        Set<IResource> partiallySharedResources = partiallySharedResourceMapping
+            .get(referencePoint);
+
+        partiallySharedResources.removeAll(resources);
+    }
+
+    /**
+     * Atomically removes and adds resources. The resources to remove will be
+     * removed first before the resources to add will be added.
+     * 
+     * @param referencePoint
+     *            a reference point that was added as a partially shared
+     *            reference point
+     * @param resourcesToRemove
+     *            the resources to remove
+     * @param resourcesToAdd
+     *            the resources to add
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
+     * 
+     * @throws IllegalStateException if the referencePoint is completely or not
+     * shared at all
+     */
+    public synchronized void removeAndAddResources(
+        IReferencePoint referencePoint,
+        Collection<? extends IResource> resourcesToRemove,
+        Collection<? extends IResource> resourcesToAdd) {
+
+        removeResources(referencePoint, resourcesToRemove);
+        addResources(referencePoint, resourcesToAdd);
+    }
+
+    /**
+     * Returns the ID assigned to the given shared reference point.
+     * 
+     * @param referencePoint
+     *            the reference point to look up the ID for
+     * @return the shared reference point's ID or <code>null</code> if the
+     *         reference point is not shared
+     */
+    public synchronized String getID(IReferencePoint referencePoint) {
+        return referencePointToIDMapping.get(referencePoint);
+    }
+
+    /**
+     * Returns the shared reference point with the given ID.
+     * 
+     * @param id
+     *            the ID to look up the reference point for
+     * @return the shared reference point for the given ID or <code>null</code>
+     *         if no shared reference point is registered with this ID
+     */
+    public synchronized IReferencePoint getReferencePoint(String id) {
+        return idToReferencePointMapping.get(id);
+    }
+
+    /**
+     * Returns whether the given resource is included in one of the currently
+     * shared reference points.
+     * 
+     * @param resource
+     *            the resource to check for
+     * @return <code>true</code> if the resource is shared, <code>false</code>
+     *         otherwise
+     */
+    public synchronized boolean isShared(IResource resource) {
+        if (resource == null)
+            return false;
+
+        if (resource.getType() == IResource.PROJECT) {
+            return isShared(resource.getReferencePoint());
+        }
+
+        IReferencePoint referencePoint = resource.getReferencePoint();
+
+        if (!idToReferencePointMapping.containsValue(referencePoint))
+            return false;
+
+        if (isCompletelyShared(referencePoint))
+            // TODO how should partial sharing handle this case ?
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=211&issues=AWVYURY9iZQY_aXE9GJQ&open=AWVYURY9iZQY_aXE9GJQ)

[ISSUE_KEY]: AWVYURY9iZQY_aXE9GJQ